theseus/geometry/functions/so3_function.py

```diff
@@ -0,0 +1,202 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+import theseus
+
+from typing import Optional, List, cast
+
+from .lie_group_function import LieGroupExpMap
+from .utils import check_jacobians_list
+
+
+def name() -> str:
+    return "SO3"
+
+
+def dim() -> int:
+    return 3
+
+
+def check_group_tensor(tensor: torch.Tensor) -> bool:
+    with torch.no_grad():
+        if tensor.ndim != 3 or tensor.shape[1:] != (3, 3):
+            raise ValueError("SO3 data tensors can only be 3x3 matrices.")
+
+        MATRIX_EPS = theseus.constants._SO3_MATRIX_EPS[tensor.dtype]
+        if tensor.dtype != torch.float64:
+            tensor = tensor.double()
+
+        _check = (
+            torch.matmul(tensor, tensor.transpose(1, 2))
+            - torch.eye(3, 3, dtype=tensor.dtype, device=tensor.device)
+        ).abs().max().item() < MATRIX_EPS
+        _check &= (torch.linalg.det(tensor) - 1).abs().max().item() < MATRIX_EPS
+
+    return _check
+
+
+def check_tangent_vector(tangent_vector: torch.Tensor) -> bool:
+    _check = tangent_vector.ndim == 3 and tangent_vector.shape[1:] == (3, 1)
+    _check |= tangent_vector.ndim == 2 and tangent_vector.shape[1] == 3
+    return _check
+
+
+class ExpMap(LieGroupExpMap):
+    @classmethod
+    def call(
```

Comment:
The list of jacobians is something we did in `theseus` as a way to avoid having multiple return types for Lie Group operations, so that we could do things like `x.compose(x.inverse())`, even if we wanted to return the jacobians. For functions with `forward()`, and `backward()` we don't need this syntax anymore, which always felt a bit clunky. 

I would prefer a structure like this:

```python
def _so3_exp_map_impl(tangent_vector: torch.Tensor) -> Tuple[torch.Tensor, ...]:
    # Returns the matrix result, as well as any intermediate tensor needed to compute the jacobian

def _so3_jac_exp_map_impl(tangent_vector) -> Tuple[torch.Tensor, torch.Tensor]:
    exp_map, other_tensors = _so3_exp_map_impl(tangent_vector)
    return _so3_jac_exp_map_aux(exp_map, other_tensors)

def _so3_jac_exp_map_aux(exp_map: torch.Tensor, *other_tensors) -> torch.Tensor:
   # Returns the jacobian given the exp_map and any pre-computed intermediate tensors 

class ExpMap(LieGroupExpMap):
    @staticmethod
    def forward(ctx, tangent_vector):
        exp_map, other_tensors = _so3_jac_exp_map_impl(tangent_vector)
        ctx.save_for_backward(exp_map, *other_tensors)

   @staticmethod
   def backward(ctx, grad_output):
        # Compute jacobian, use a cache, multiply by grad_output, etc. 
        # Here we just call _so3_jac_exp_map_aux(ctx.exp_map, *ctx.other_tensors)
```

Of course, the function to compute the jacobians can also be public, since we'll probably use it in some cost functions.

Reply:
Thanks for the detailed structure. If I understand correctly, jacobians should be one of the returns of `forward()`. Thus, we have to provide an analytical `backward()` for the jacobians as well, which can be very complicated.

Reply:
Do you mean in my proposal? If so, then no, I wasn't suggesting jacobians are returned by forward, only the exp map. But I just noticed there was a typo, the first line in my forward was supposed to be 
```python
exp_map, other_tensors = _so3_exp_map_impl(tangent_vector)
```
The jacobians are only computed in backward.

Reply:
Got it. But the `jacobians` got from `backward()` are `Euclidean gradients`. How can people get access to `Lie tangent jacobians`?

Reply:
Could be a separate function inside so3.py like
```python
def jexp_tangent(tangent_vector: torch.Tensor) -> torch.Tensor:
    ...
```
or if the logic allows it, we could do
```python
def jexp(tangent_vector: torch.Tensor, is_euclidean: bool = False) -> torch.Tensor:
     pass
```
BTW, I'm thinking that maybe we don't need the word "map" for log map, or exp map? I think short names would make sense to be in the spirit of a lightweight library.

Reply:
I would like the jacobians to be computed with `exp_map`, which some of the computation can be reused.

Reply:
I already proposed one way to reuse computation, by returning intermediate tensors as a separate output in the private `_exp_map_impl`. This lets you defer the final computatio of the jacobians to only be done if backward is called.

Reply:
I agree that we want to move away from the cpp like calls that fill jacobians passed as a reference in the function call.

For a python/pytorch user we need to support say
```
y = x.compose(x.inverse())
y.backward()
```

But we also need the ability to get function jacobians for manual cost functions (not based on AutoDiffCostFunction). Would we just call a partial backward here? For this to be efficient maybe we need to support both forward and backward mode differentiation?